crates/order-validation/src/banned.rs

```diff
@@ -1,34 +1,183 @@
 use {
-    cached::{Cached, TimedCache},
     contracts::ChainalysisOracle,
-    ethcontract::{errors::MethodError, futures::future::join_all, H160},
-    std::{collections::HashSet, sync::Mutex},
+    ethcontract::{
+        errors::MethodError,
+        futures::{future::join_all, TryStreamExt},
+        H160,
+    },
+    sqlx::PgPool,
+    std::{
+        collections::{HashMap, HashSet},
+        ops::Div,
+        sync::Arc,
+        time::{Duration, Instant},
+    },
+    tokio::sync::RwLock,
 };
 
 /// A list of banned users and an optional registry that can be checked onchain.
 pub struct Users {
     list: HashSet<H160>,
-    onchain: Option<Onchain>,
+    onchain: Option<Arc<Onchain>>,
+}
+
+#[derive(Clone)]
+struct UserMetadata {
+    is_banned: bool,
+    last_updated: Instant,
+    limit_order_participant: bool,
+}
+
+impl UserMetadata {
+    pub fn with_banned(mut self, banned: bool) -> Self {
+        self.is_banned = banned;
+        self
+    }
+
+    pub fn with_last_updated(mut self, last_updated: Instant) -> Self {
+        self.last_updated = last_updated;
+        self
+    }
 }
 
 struct Onchain {
     contract: ChainalysisOracle,
-    cache: Mutex<TimedCache<H160, bool>>,
+    db_pool: PgPool,
```

Comment:
To get the list of owners and receivers from the current auction it seems more straight forward to have a reference to the `SolvableOrdersCache` here if possible instead of making DB requests.

Reply:
I don't think it is possible since this crate is used in both autopilot and the orderbook. Extracting the SolvableOrdersCache from the autopilot to another crate seems too much overhead. 

I really wanted to avoid fetching auction JSON, but it seems like this is the only way.

Reply:
> used in both autopilot and the orderbook

Mmmh, good point. I think the prefetching (actually any cache maintenance at all) only really makes sense in the autopilot case since it would be fine for users to wait 1s for us to check IMO.
Another thing that might work is to provide a cache eviction function in the constructor. That way you could pass a closure that has a reference to the `SolvableOrdersCache` in the `autopilot` or a dummy implementation that evicts all expired addresses in the orderbook.

Reply:
Due to the circular dependency between `infra::banned::Users` and `SolvableOrdersCache`, that would probably be tricky to find a concise solution...

Reply:
Unless I'm missing something this circular dependency will not result in deadlocks because the `SolvableOrdersCache` only locks AFTER the banned user filtering ran.